src/demos/findex/findex.ts

```diff
@@ -0,0 +1,186 @@
+import { AbeMasterKey } from "../../crypto/abe/keygen/keygen";
+import { Policy, PolicyAxis } from "../../crypto/abe/keygen/policy";
+import { generateMasterKeys, coverCryptEncrypt, coverCryptDecrypt } from "../abe/cover_crypt/cover_crypt";
+import { Findex } from "../../interface/findex/findex";
+import { DB } from "./demo_db";
+import { masterKeysFindex } from "./demo_keys";
+import { logger } from "../../utils/logger";
+import { hexDecode, hexEncode, sanitizeString } from "../../utils/utils";
+
+export class FindexDemo {
+  private _db: DB;
+  private _policy: Policy;
+  private _masterKeysCoverCrypt: AbeMasterKey;
+
+  constructor(db: DB, policyAxes: PolicyAxis[], maxAttributeCreations: number) {
+    this._db = db;
+    this._policy = new Policy(policyAxes, maxAttributeCreations)
+    this._masterKeysCoverCrypt = generateMasterKeys(this._policy);
+  }
+
+  public get db(): DB {
+    return this._db;
+  }
+  public get policy(): Policy {
+    return this._policy;
+  }
+  public get masterKeysCoverCrypt(): AbeMasterKey {
+    return this._masterKeysCoverCrypt;
+  }
+
+  async encryptUsers(metadataUid: Uint8Array) {
+    const policyBytes = this._policy.toJsonEncoded();
+
+    const users = await this._db.getUsers();
+    for (const user of users) {
+      const encryptedBasic = coverCryptEncrypt(
+        policyBytes,
+        this._masterKeysCoverCrypt.publicKey,
+        metadataUid,
+        [`department::marketing`, `country::${user.country}`],
+        JSON.stringify({
+          firstName: user.firstName,
+          lastName: user.lastName,
+          country: user.country,
+          region: user.region
+        }))
+
+      const encryptedHr = coverCryptEncrypt(
+        policyBytes,
+        this._masterKeysCoverCrypt.publicKey,
+        metadataUid,
+        [`department::HR`, `country::${user.country}`], JSON.stringify({
+          email: user.email,
+          phone: user.phone,
+          employeeNumber: user.employeeNumber
+        })
+      )
+
+      const encryptedSecurity = coverCryptEncrypt(
+        policyBytes,
+        this._masterKeysCoverCrypt.publicKey,
+        metadataUid,
+        [`department::security`, `country::${user.country}`], JSON.stringify({
+          security: user.security
+        }))
+
+      const upsertedEncElement = await this._db.upsertEncryptedUser({
+        enc_basic: hexEncode(encryptedBasic),
+        enc_hr: hexEncode(encryptedHr),
+        enc_security: hexEncode(encryptedSecurity)
+      });
+
+      await this._db.upsertUserEncUidById(user.id, { enc_uid: upsertedEncElement[0].uid });
+    };
+
+  }
+
+  /**
+   * Reset all indexes and upsert new ones
+   *
+   * @param location location string naming the key of location to index
+   */
+  async resetAndUpsert(location: string) {
```

Comment:
You call this function with `enc_uid` but it seems that the SQL request in `getUsers()` doesn't contains this key…

I propose to create a real type `User` inside `demo_db.ts` (it will avoid repetition of the attributes)
```
export interface User {
  id: string, firstName: string, lastName: string, phone: string, email: string, country: string, region: string, employeeNumber: string, security: string,
}
```

Here you can then import this new type, remove your fake `type User` and `location` can be of type `keyof User` (add some compiler check)